ArduCopter/mode_autorotate.cpp

```diff
@@ -85,24 +85,62 @@ void ModeAutorotate::run()
     // Current time
     uint32_t now = millis(); //milliseconds
 
+    // set dt in library
+    float const last_loop_time_s = AP::scheduler().get_last_loop_time_s();
+    g2.arot.set_dt(last_loop_time_s);
+
+    float alt = g2.arot.get_ground_distance();
+    // have autorotation library update estimated radar altitude
+    g2.arot.update_est_rangefinder_alt();
+    if (alt < copter.rangefinder.max_distance_cm_orient(ROTATION_PITCH_270)) {
+        g2.arot._using_rfnd = true;
+    } else {
+        g2.arot._using_rfnd = false;
+    }
+
     // Initialise internal variables
     float curr_vel_z = inertial_nav.get_velocity_z_up_cms();   // Current vertical descent
+    //calculate time to impact
+    float time_to_impact;
+    if (phase_switch != Autorotation_Phase::TOUCH_DOWN) {
+        g2.arot.time_to_ground();
+        time_to_impact = g2.arot.get_time_to_ground();
+        last_tti=time_to_impact;
+    } else {
+        time_to_impact = last_tti;
+    }
+
 
     //----------------------------------------------------------------
     //                  State machine logic
     //----------------------------------------------------------------
 
-     // Setting default phase switch positions
-     nav_pos_switch = Navigation_Decision::USER_CONTROL_STABILISED;
-
-    // Timer from entry phase to progress to glide phase
-    if (phase_switch == Autorotation_Phase::ENTRY){
-
-        if ((now - _entry_time_start_ms)/1000.0f > AUTOROTATE_ENTRY_TIME) {
-            // Flight phase can be progressed to steady state glide
-            phase_switch = Autorotation_Phase::SS_GLIDE;
+    if (initial_energy_check) {
+        //initial check for total energy monitoring
+        if ( inertial_nav.get_speed_xy_cms() < 250.0f && g2.arot.get_ground_distance() < g2.arot.get_flare_alt()) {
+            hover_autorotation = true;
```

Comment:
From what I can gather, the key difference in the hover_autorotation is that it when between the flare height and touch down phase it drives the collective to mid col (I am guessing to minimise rotor drag) until we get into the touch down time to impact and then we advance to the touch down phase.  @Ferruccio1984 can you confirm that I have understood this correctly?  If so, my preference is that we make the hover autororation an additional Autorotation_Phase.  It will make the code easier to read (and therefore less error prone) as it is a little convoluted having a special case within the state machine.

Reply:
> can you confirm that I have understood this correctly?

@Gone4Dirt that's correct, this was introduced after a test case showed the user was hovering at 10m AGL and performed an hover auto with an obvious heavy landing. Saving a little bit of rotor energy might help for those cases where it is difficult to assess what the optimal altitude would be for testing hover autos. Also, this is something that will have to be mentioned in the wiki...HIGE :-) !